packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -609,23 +602,24 @@ function initializeMode(vm: VM) {
         );
     }
 
+    // Default modes.
+    vm.renderMode = RenderMode.Shadow;
+    vm.shadowMode = ShadowMode.Native;
+
     if (renderMode === 'light') {
         vm.renderMode = RenderMode.Light;
-        vm.shadowMode = ShadowMode.None;
-        return;
+    } else if (renderer.ssr) {
+        vm.renderMode = RenderMode.Shadow;
     }
 
     if (renderer.ssr) {
-        vm.renderMode = RenderMode.Shadow;
         vm.shadowMode = ShadowMode.Native;
-        return;
+    } else if (renderer.syntheticShadow) {
+        vm.shadowMode =
+            isTrue(preferNativeShadow) && isNativeShadowRootDefined
```

Comment:
I'm still fuzzy about `isNativeShadowRootDefined`, and how is that set? I always thought that this was backward, in the sense that you might detect when the synthatic shadow is available. I suspect this might be a gate/perm related stuff?

Reply:
We added it relatively recently [here](https://github.com/salesforce/lwc/blob/75d2bcccb7de4d4c9fc93d8fb41aaf282bb6f543/packages/%40lwc/synthetic-shadow/src/env/shadow-root.ts). It's set by synthetic shadow and is `undefined` when synthetic shadow is not loaded.